src/Core/Ast/AstMap/TaggedReferenceTrait.php

```diff
@@ -0,0 +1,38 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Qossmic\Deptrac\Core\Ast\AstMap;
+
+/**
+ * Helper trait for implementing TaggedTokenReferenceInterface.
+ *
+ * Classes that use this trait must define $this->tags as array<string,string[]>.
+ *
+ * @psalm-immutable
+ */
+trait TaggedReferenceTrait
```

Comment:
Why did you decide to use an interface and a trait? Why not `abstract class` instead? It would allow you to enforce the `$this->tags` property on the child classes and you would not have to mention in in the doc comments.

Reply:
I didn't want to base the ontology of lexical entities on the question whether they can have tags or not. For instance, I could imagine a future where we'd want a FunctionLikeReference, with subclasses MethodReference, FunctionReference, and ClosureReference. MethodsReference and FunctionReference would be taggable, ClosureReference would not.

But perhaps I am being overly cautious. From my experience in MediaWiki,  I have become very careful about class hierarchies, since they tend to be very hard to change in a backwards compatible way. But perhaps that concern is misplaced here. Deptrac doesn't expose as much of its internals to extendsions as MediaWiki does. Though I can still see extensions type hinting against base classes of TokenReferences.

Anyway, I agree that way fields work in traits is rather unfortunate. I wish traits had constructors.

Reply:
>Though I can still see extensions type hinting against base classes of TokenReferences.

This is intentionally not possible/allowed since the base classes are not part of the `Contract` layer but safely tucked away in `Ast` layer that is not exposed to the public for writing extensions. Therefore you don't have to worry about backwards compatibility.

Reply:
Does that mean that extensions would not be able to type-hint against a TaggedTokenReference base class?
I suppose we'd still need the TaggedTokenReferenceInterface, even if we used a base class instead of a trait?

Reply:
It depends on where you put it. I would put it into the `Contract` alongside `TokenReferenceInterface`. I see no issue having an `abstract class TakkedTokenReference` in the `Contract` with the `tags` property and implementation of the getter methods you currently have in the `TaggedTokenReferenceInterface`.

Reply:
I converted the trait into a base class, but kept it internal in Core, and kept the interface in Contract.
If we exposed the base class in Contract it's no longer an implementation detail that can easily be changed later.